iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+            / pl.std("estimate"),
+        )
+
+        x = (
+            train_std.select(["previous_std", "elapsed_std"])
+            .with_columns(interact=pl.col("previous_std") * pl.col("elapsed_std"))
+            .to_numpy()
+        )
+        y = train_std.select(["estimate_std"]).to_numpy()
+
+        self.model.fit(x, y)
+
+        return self
+
+    def predict(
+        self,
+        train: IncidentUptakeData,
+        start_date,
+        end_date,
+        interval,
+        init_elapsed,
+        init_interval,
+        init_incident,
+        init_cumulative,
+    ) -> Self:
+        previous_mn = train["previous"].mean()
+        previous_sd = train["previous"].std()
+        elapsed_mn = train["elapsed"].mean()
+        elapsed_sd = train["elapsed"].std()
+        estimate_mn = train["estimate"].mean()
+        estimate_sd = train["estimate"].std()
+
+        incident_proj = (
+            pl.date_range(
+                start=start_date,
+                end=end_date,
+                interval=interval,
+                eager=True,
+            )
+            .alias("date")
+            .to_frame()
+            .with_columns(
+                region=pl.lit(train.select(pl.col("region").unique())),
+                elapsed=(
+                    (pl.col("date") - start_date).cast(pl.Float64)
+                    / (1000 * 60 * 60 * 24)
+                )
+                + init_elapsed,
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
+            )
+            .with_columns(
+                interval=pl.col("elapsed").diff().fill_null(init_interval),
+                season=pl.col("season").cast(pl.Utf8),
+            )
+        )
+
+        proj = np.zeros((incident_proj.shape[0]))
+        proj[0] = init_incident
+
+        for i in range(proj.shape[0] - 1):
+            x = np.column_stack(
+                (
+                    (proj[i] - previous_mn) / previous_sd,
+                    (incident_proj["elapsed"][i + 1] - elapsed_mn) / elapsed_sd,
+                )
+            )
+            x = np.insert(x, 2, np.array((x[:, 0] * x[:, 1])), axis=1)
+            y = self.model.predict(x)
+            proj[i + 1] = y[(0, 0)] * estimate_sd + estimate_mn
+
+        incident_proj = incident_proj.with_columns(estimate=pl.Series(proj))
+
+        cumulative_proj = incident_proj.with_columns(
+            estimate=pl.col("estimate") * pl.col("interval")
+        ).with_columns(estimate=pl.col("estimate").cum_sum() + init_cumulative)
+
+        incident_proj = IncidentUptakeProjection(
+            incident_proj.select("region", "date", "estimate", "interval")
+        )
+        incident_proj.validate()
+
+        cumulative_proj = CumulativeUptakeProjection(
+            cumulative_proj.select("region", "date", "estimate")
+        )
+        cumulative_proj.validate()
+
+        setattr(self, "incident_projections", incident_proj)
```

Comment:
What's going on here?

I *think* you're doing something that, in python, is `self.incident_projections = indicent_proj`

But while fitting is something that modifies the object itself, and gets stored in its state, predictions are not. This model should output one thing (presumably incident uptake) and just return it, without modifying self

Reply:
Because each LinearIncidentUptakeModel (for which model structure is fixed) belongs to an instance of ProjectionSettings (for which the training data and test time-frame is fixed), it produces exactly one set of projections. So I figured these projections (both their incident and cumulative versions) might as well belong to the LinearIncidentUptakeModel as attributes.

Reply:
mmmm good question about how we want this to work

The scikitlearn paradigm is: that you pick a model type (e.g., logistic regression, or random forest), you instantiate a model object (with whatever hyperparameters you want in there), then you `.train()` that model object (changing its state), and `.predict()` (producing output but not changing state). Then you can take the same model object and train it on different data.

So the equivalent here is would be instantiating one model object (e.g., that does linear uptake projections), and then for each forecast date, retrain the model (changing its state) and doing a prediction (not changing its state)

I'm not sure if this paradigm is the best one, but I do like the separation of hyperparameterization (no hyperparameters in a linear regression, but in random forest you would pick the number of trees, say), training (which fits the parameters), and prediction.

cc @afmagee42 if you have other ideas, or if there's some learning from variant nowcasting that I've missed here

Reply:
We've actually been considering caching the predictions within the model object for variant nowcasting.

This is partially because they can be very expensive to generate (we're talking about drawing from 52 x 105 = 5460 roughly 10-D multinomials each with sample sizes easily in the hundreds _per posterior sample_, that's billions of random numbers).

And it's partially because we've been thinking of moving away from passing outputs around to passing objects (see https://github.com/CDCgov/cfa-viral-lineage-model/issues/52). So, you instantiate a `Model` object with data, then `model.fit()` returns a `FitModel` which can be passed to evaluators and such. In this case, you basically have to store the predictions.

(I agree that `setattr` isn't very "pythonic" and the usual style is `self.thing = value`.)

Reply:
And I'm not personally sure I see a ton of value in being able to instantiate a hyperparameter-specified, data-free model? You can't do that in `stan`, FWIW.

Reply:
I think the value is something that feels like partial application:

- Model class is very general
- Model object has hyperparameters but no data
- Fit model object (which, in the scikit-learn paradigm, is the same object, just with a different state) has data and posterior
- Fit model object can predict, which doesn't need to change state of anything (although it makes good sense to cache predictions inside the object, and clear the cache if you refit)

I take your arrangement as:

- Model class is very general
- Model object has parameters *and* data
- `my_model.fit()` returns a FitModel object (of a different class)
- `my_fit_model.predict()`

I think the appeal of the scikit-learn arrangement is that you could re-train your `Model`, or you could copy it and fit it on multiple data sets, and none of the functions that do that later manipulation need to remember what the hyperparameters were

Reply:
While I don't see the value of "retrain your model," I _do_ see an ease-of-use argument for not constantly having to set the same hyperparameters repeatedly if repeatedly fitting the same model to different datasets.

I think I could be happy with a "same object, different state" paradigm since in theory many flavors of Bayesian model can `predict()` from the prior as a default. As long as something can be passed around, possibly modified as needed, and asked for predictions.

Reply:
We could always add to the levels of hierarchy. `Model` (abstract-ish) -> `TunedModel` (has hyperparameters, MCMC settings, etc) -> `FittedModel`